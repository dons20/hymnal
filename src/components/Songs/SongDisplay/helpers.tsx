import React from 'react';
import { Box, Text } from '@mantine/core';

/**
 * Helper function to render chorus with vocal parts and indicators
 */
export const renderChorusWithParts = (song: Song, isDark: boolean) => {
  if (!song.chorusParts || song.chorusParts.length === 0) {
    // Fallback to regular chorus rendering
    return <Text className="song-text">{song.chorus}</Text>;
  }

  return (
    <Box className="chorus-parts">
      {song.chorusParts.map((part, index) => (
        <Box
          key={index}
          className="chorus-part"
          style={{ position: 'relative', marginBottom: '0.5rem' }}
        >
          {part.part && (
            <Text
              component="span"
              className="part-indicator-regular"
              style={
                {
                  '--part-indicator-color':
                    part.part === 'soprano'
                      ? isDark
                        ? '#74c0fc'
                        : '#1c7ed6' // Blue for soprano
                      : isDark
                        ? '#ffec99'
                        : '#fd7e14', // Orange for baritone
                  '--part-indicator-bg':
                    part.part === 'soprano'
                      ? isDark
                        ? 'rgba(116, 192, 252, 0.15)'
                        : 'rgba(28, 126, 214, 0.08)'
                      : isDark
                        ? 'rgba(255, 236, 153, 0.15)'
                        : 'rgba(253, 126, 20, 0.08)',
                  '--part-indicator-border':
                    part.part === 'soprano'
                      ? isDark
                        ? 'rgba(116, 192, 252, 0.3)'
                        : 'rgba(28, 126, 214, 0.25)'
                      : isDark
                        ? 'rgba(255, 236, 153, 0.3)'
                        : 'rgba(253, 126, 20, 0.25)',
                } as React.CSSProperties
              }
            >
              {part.part === 'soprano' ? 'A' : part.part === 'baritone' ? 'B' : '♪'}
            </Text>
          )}
          <Text
            className="song-text"
            style={
              {
                paddingLeft: part.part ? '0' : '0',
                fontWeight: part.part === 'soprano' ? 500 : 400,
                '--part-text-color':
                  part.part === 'soprano'
                    ? isDark
                      ? '#e9ecef'
                      : '#212529' // Standard text color for soprano
                    : isDark
                      ? '#ced4da'
                      : '#495057', // Slightly muted for baritone
                color: 'var(--part-text-color)',
              } as React.CSSProperties
            }
          >
            {part.text}
          </Text>
        </Box>
      ))}
    </Box>
  );
};

/**
 * Helper function to render presentation slide content (verses or chorus with vocal parts)
 */
export const renderPresentationContent = (content: string | VocalPart[], isDark: boolean) => {
  if (typeof content === 'string') {
    return <Text className="presentation-text">{content}</Text>;
  }

  // Handle VocalPart[] for presentation mode with larger styling
  return (
    <Box className="presentation-content-container">
      {content.map((part, index) => {
        // Check if this is the end of a B part followed by an A part (new pair starting)
        const isEndOfPair =
          part.part === 'baritone' &&
          index < content.length - 1 &&
          content[index + 1]?.part === 'soprano';

        return (
          <React.Fragment key={index}>
            <Box
              className="presentation-part-container"
              style={{
                marginBottom: isEndOfPair ? '2rem' : '1.5rem',
              }}
            >
              {part.part && (
                <Text
                  component="span"
                  className="part-indicator-presentation"
                  style={
                    {
                      '--part-indicator-color':
                        part.part === 'soprano'
                          ? isDark
                            ? '#74c0fc'
                            : '#1c7ed6' // Blue for soprano
                          : isDark
                            ? '#ffec99'
                            : '#fd7e14', // Orange for baritone
                      '--part-indicator-bg':
                        part.part === 'soprano'
                          ? isDark
                            ? 'rgba(116, 192, 252, 0.15)'
                            : 'rgba(28, 126, 214, 0.08)'
                          : isDark
                            ? 'rgba(255, 236, 153, 0.15)'
                            : 'rgba(253, 126, 20, 0.08)',
                      '--part-indicator-border':
                        part.part === 'soprano'
                          ? isDark
                            ? 'rgba(116, 192, 252, 0.3)'
                            : 'rgba(28, 126, 214, 0.25)'
                          : isDark
                            ? 'rgba(255, 236, 153, 0.3)'
                            : 'rgba(253, 126, 20, 0.25)',
                    } as React.CSSProperties
                  }
                >
                  {part.part === 'soprano' ? 'A' : part.part === 'baritone' ? 'B' : '♪'}
                </Text>
              )}
              <Text
                className="presentation-part-text"
                style={
                  {
                    fontWeight: part.part === 'soprano' ? 600 : 400,
                    '--part-text-color':
                      part.part === 'soprano'
                        ? isDark
                          ? '#e9ecef'
                          : '#212529' // Standard text color for soprano
                        : isDark
                          ? '#ced4da'
                          : '#495057', // Slightly muted for baritone
                  } as React.CSSProperties
                }
              >
                {part.text}
              </Text>
            </Box>
            {isEndOfPair && (
              <Box
                className="pair-divider"
                style={
                  {
                    '--divider-color': isDark
                      ? 'rgba(255, 255, 255, 0.2)'
                      : 'rgba(0, 0, 0, 0.15)',
                  } as React.CSSProperties
                }
              />
            )}
          </React.Fragment>
        );
      })}
    </Box>
  );
};
