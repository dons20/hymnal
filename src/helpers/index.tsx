/* eslint-disable no-console */
import React from "react";
import axios from "axios";
import hymns from "static/hymns.json";
import setDebug from "utils/logger";

interface SongPropsA {
	number: number;
	title: string;
	verse: string[];
	chorus: string;
	author?: string;
}
interface SongPropsB {
	number: number;
	title: string;
	verse: string[];
	chorus: string;
	author: string;
}

export type SongProps = SongPropsA | SongPropsB;

export type SongsDB = { songs: SongProps[]; version: string };

setDebug();
const { debug } = window;

/**
 * Loads songs from JSON and stores them locally
 */
export async function loadNewSongs(fetchedSongs: SongProps[]) {
	const localForage = await import("localforage");
	const songs = localForage.createInstance({ storeName: "items" });
	const localVersion = localForage.createInstance({ storeName: "version" });
	const favourites = localForage.createInstance({ storeName: "favourites" });

	try {
		await Promise.all([
			fetchedSongs.forEach((song, i) => songs.setItem(`${i}`, { ...song }).catch(e => debug.info(e))),
			localVersion.setItem("value", localVersion).catch(e => debug.info(e)),
			favourites.clear(),
		]);
	} catch (err) {
		if (err instanceof Error) debug.info(err.message);
	}
}

/**
 * Checks if songs have already been stored
 */
export async function checkDB() {
	const dbName = "Songs";
	const localForage = await import("localforage");
	localForage.config({
		name: dbName,
		description: "Stores the songs db and its version number",
	});

	let songsDB: SongsDB;
	try {
		const fetchedData = process.env.NODE_ENV === "production" ? await axios.get<SongsDB>(process.env.HYMNS_URL || "https://f002.backblazeb2.com/file/hymnal/hymns.json") : { data: hymns };
		if (!fetchedData.data) return;
		songsDB = fetchedData.data;

		debug.log(`%cChecking if songs exist already`, "color: #3182ce; font-size: medium;");
		const songs = localForage.createInstance({ storeName: "items" });
		const songsLength = await songs.length();
		if (songsLength < songsDB.songs.length) {
			debug.info("Items out of sync with latest items");
			loadNewSongs(songsDB.songs);
		}

		debug.log(`%cChecking for updates`, "color: #3182ce; font-size: medium;");
		const localVersion = localForage.createInstance({ storeName: "version" });
		if (!localVersion) throw new Error("No version stored");
		const versionNumber = (await localVersion.getItem("value")) as string;
		if (songsDB.version !== versionNumber) {
			debug.info("Version mismatch, sync necessary");
			loadNewSongs(songsDB.songs);
		}

		debug.log(`%cSongs found! Attempting to load...`, "color: #3182ce; font-size: medium;");
	} catch (e) {
		if (e instanceof Error) debug.error(e.message);
	}
}

export function useQuery(search: string) {
	return new URLSearchParams(search);
}

/**
 * A helper to create a Context and Provider with no upfront default value, and
 * without having to check for undefined all the time.
 */
export function createCtx<A extends {} | null>() {
	const ctx = React.createContext<A | undefined>(undefined);
	function useCtx() {
		const c = React.useContext(ctx);
		if (c === undefined) throw new Error("useCtx must be inside a Provider with a value");
		return c;
	}
	return [useCtx, ctx.Provider, ctx.Consumer] as const; // 'as const' makes TypeScript infer a tuple
}

export async function updateFavesDB(indexes: number[]) {
	const name = "Songs";
	const storeName = "Favourites";
	const localForage = await import("localforage");

	localForage.config({
		name,
		storeName,
		description: "Your favourite songs",
	});

	indexes.forEach(async item => {
		await localForage.setItem(`${item}`, item);
	});
}
