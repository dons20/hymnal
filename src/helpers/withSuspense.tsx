/* eslint-disable react/function-component-definition */
import { lazy, Suspense } from "react";

const Loader = lazy(() => import("components/Loader"));

const withSuspense =
	<P extends React.FC | Function, Q = void>(
		// @ts-ignore TODO: Need to find a way to fix the return types here
		LazyComponent: React.JSXElementConstructor<P> | React.LazyExoticComponent<P> | P,
		FallbackComponent?: React.FC<Q> | null
	): P =>
	// @ts-ignore TODO: Possibly investigate fix for this error
	(props: P, props2: Q) => {
		// eslint-disable-next-line react/jsx-no-useless-fragment
		let fallbackLoader = <></>;
		if (FallbackComponent === undefined)
			fallbackLoader = (
				<Suspense fallback="">
					<Loader />
				</Suspense>
			);
			// @ts-expect-error TODO: Fix this
		if (FallbackComponent) fallbackLoader = <FallbackComponent {...props2} />;

		return (
			<Suspense fallback={fallbackLoader}>
				{/** @ts-ignore TODO: Need to find a way to fix the return types here */}
				<LazyComponent {...props} />
			</Suspense>
		);
	};

export default withSuspense;
