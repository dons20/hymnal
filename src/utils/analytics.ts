/**
 * Analytics utilities for managing Umami tracking based on online/offline status
 */

import setDebug from './logger';

// Initialize debug logger
setDebug();
const { debug } = window;

// Extend the Window interface to include umami
declare global {
  interface Window {
    umami?: {
      track: (eventName: string, eventData?: Record<string, any>) => void;
    };
  }
}

// Store original umami functions to restore later
let originalUmami: typeof window.umami | null = null;
let isAnalyticsDisabled = false;

/**
 * Disables Umami analytics by replacing the umami object with no-op functions
 */
export function disableAnalytics() {
  if (typeof window !== 'undefined' && window.umami && !isAnalyticsDisabled) {
    debug.log('📊 Analytics disabled (offline mode)');
    
    // Store the original umami object
    originalUmami = window.umami;
    
    // Replace with no-op functions
    window.umami = {
      track: () => {
        debug.log('📊 Analytics tracking blocked (offline)');
      }
    };
    
    isAnalyticsDisabled = true;
  }
}

/**
 * Enables Umami analytics by restoring the original umami object
 */
export function enableAnalytics() {
  if (typeof window !== 'undefined' && originalUmami && isAnalyticsDisabled) {
    debug.log('📊 Analytics enabled (online mode)');
    
    // Restore the original umami object
    window.umami = originalUmami;
    originalUmami = null;
    isAnalyticsDisabled = false;
  }
}

/**
 * Manages analytics based on online/offline status
 */
export function manageAnalytics() {
  if (typeof window === 'undefined') return;
  
  if (navigator.onLine) {
    enableAnalytics();
  } else {
    disableAnalytics();
  }
}

/**
 * Sets up listeners for online/offline events to automatically manage analytics
 */
export function setupAnalyticsListeners() {
  if (typeof window === 'undefined') return;
  
  // Handle online/offline events
  window.addEventListener('online', () => {
    debug.log('🌐 Network connection restored');
    enableAnalytics();
  });
  
  window.addEventListener('offline', () => {
    debug.log('🌐 Network connection lost');
    disableAnalytics();
  });
  
  // Initial setup based on current status
  manageAnalytics();
}