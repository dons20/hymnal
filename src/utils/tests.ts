import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

export const renderWithRouter = (
	ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
	{ route = "/" } = {}
) => {
	window.history.pushState({}, "Test page", route);

	// @ts-ignore
	return render(ui, { wrapper: BrowserRouter });
};

export default renderWithRouter;
