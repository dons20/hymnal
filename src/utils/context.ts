import { createCtx } from "helpers";

export const pages = {
	HOME: "/home",
	INDEX: "/songs",
	FAVOURITES: "/favourites",
	SETTINGS: "/settings",
};

export type State = {
	title: string;
	subtitle: string;
	width: number;
};

export type ACTIONTYPE =
	| { type: "setTitle"; payload: string }
	| { type: "setSubtitle"; payload: string }
	| { type: "setWidth"; payload: number };

export type CTX = {
	dispatch: React.Dispatch<ACTIONTYPE>;
	songs: Song[];
	favourites: number[];
	setFavourites: React.Dispatch<React.SetStateAction<number[]>>;
	pages: typeof pages;
	meta: State;
};

export const [useMainContext, MainContextProvider, MainContextConsumer] = createCtx<CTX>();
