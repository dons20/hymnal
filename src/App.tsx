import { useReducer, useEffect, lazy } from "react";
import { Route, Switch, Redirect, useLocation } from "react-router-dom";
import { useColorModeValue } from "@chakra-ui/color-mode";
import { useSongLoader } from "components/CustomHooks";
import withSuspense from "helpers/withSuspense";
import { isMobile } from "react-device-detect";
import { Box } from "@chakra-ui/layout";

import { State, ACTIONTYPE, MainContextProvider, pages } from "utils/context";
import styles from "./App.module.scss";

const HomeImport = lazy(() => import("pages/Home"));
const SongsImport = lazy(() => import("pages/Songs"));
const SearchImport = lazy(() => import("pages/Search"));
const LoaderImport = lazy(() => import("components/Loader"));
const HeaderImport = lazy(() => import("components/Header"));
const BottomNavImport = lazy(() => import("components/BottomNav"));
const PictureHeaderImport = lazy(() => import("components/PictureHeader"));

const Home = withSuspense<typeof HomeImport>(HomeImport);
const Songs = withSuspense<typeof SongsImport>(SongsImport);
const Search = withSuspense<typeof SearchImport>(SearchImport);
const Loader = withSuspense<typeof LoaderImport, null>(LoaderImport, null);
const Header = withSuspense<typeof HeaderImport, null>(HeaderImport, null);
const BottomNav = withSuspense<typeof BottomNavImport, null>(BottomNavImport, null);
const PictureHeader = withSuspense<typeof PictureHeaderImport, null>(PictureHeaderImport, null);

const initialAppState = {
	title: "",
	subtitle: "",
	width: document.body.getBoundingClientRect().width,
};

function reducer(state: State, action: ACTIONTYPE): State {
	switch (action.type) {
		case "setTitle":
			return { ...state, title: action.payload };
		case "setSubtitle":
			return { ...state, subtitle: action.payload };
		case "setWidth":
			return { ...state, width: action.payload };
		default:
			return state;
	}
}

const ScrollRestoration = () => {
	const { pathname } = useLocation();
	useEffect(() => {
		window.scrollTo(0, 0);
	}, [pathname]);
	return null;
};

function App() {
	const [state, dispatch] = useReducer(reducer, initialAppState);
	const { songs, favourites, setFavourites } = useSongLoader();
	const pageBG = useColorModeValue("gray.200", "gray.800");

	const handleOrientationChange = () => {
		dispatch({ type: "setWidth", payload: document.body.getBoundingClientRect().width });
	};

	useEffect(() => {
		window.screen.orientation.addEventListener("change", handleOrientationChange, {
			capture: false,
			passive: true,
		});
		return function cleanup() {
			window.screen.orientation.removeEventListener("change", handleOrientationChange);
		};
	}, []);

	return (
		<MainContextProvider value={{ meta: state, dispatch, songs, pages, favourites, setFavourites }}>
			<Box className={styles.root}>
				<Box as="section" className={styles.app_body}>
					<Header />
					<ScrollRestoration />
					<Box
						as="main"
						className={`${styles.app_inner} ${isMobile ? styles.app_inner_mobile : ""}`}
						bg={pageBG}
					>
						<PictureHeader />

						<div className={styles.wrapper}>
							<Switch>
								<Route path="/home" component={Home} />
								<Route path="/songs">{songs.length > 1 ? <Songs /> : <Loader />}</Route>
								<Route path="/search" component={Search} />
								<Route path="/favourites">
									<Redirect to="/songs/favourites" />
								</Route>
								<Route>
									<Redirect to="/home" />
								</Route>
							</Switch>
						</div>
					</Box>

					<BottomNav />
				</Box>
			</Box>
		</MainContextProvider>
	);
}

export default App;
