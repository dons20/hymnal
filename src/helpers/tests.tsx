import { act } from "react-dom/test-utils";
import { MainContextProvider } from "utils/context";

type ProviderT = {
	children: React.ReactNode;
	value: any;
};

const resizeWindow = (x: number, y: number) => {
	// @ts-ignore
	window.innerWidth = x;
	// @ts-ignore
	window.innerHeight = y;
	act(() => {
		window.dispatchEvent(new Event("resize"));
	});
};

const Providers = ({ children, value }: ProviderT) => (
	<MainContextProvider value={value}>{children}</MainContextProvider>
);

export { resizeWindow, Providers };
export default resizeWindow;
