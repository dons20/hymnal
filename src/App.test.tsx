// jest.mock("localforage");
import { Suspense } from "react";
import { screen, waitFor } from "@testing-library/react";
import { renderWithRouter } from "utils/tests";
import { resizeWindow } from "helpers/tests";
import localforage from "localforage";
import App from "./App";

jest.mock("__mocks__/createMocks.ts");

beforeAll(async () => {
	const addEventListener = jest.fn();
	const removeEventListener = jest.fn();
	jest.spyOn(console, "log").mockImplementation(() => {});

	// @ts-ignore
	window.screen.orientation = { addEventListener, removeEventListener };
	window.scrollTo = jest.fn();
});

afterEach(() => {
	localforage.clear();
});

describe("#App", () => {
	it("renders without crashing", async () => {
		const spyConfig = jest.spyOn(localforage, "config");
		const spyCreateInstance = jest.spyOn(localforage, "createInstance");
		await waitFor(() => {
			expect(spyConfig).toBeTruthy();
			expect(spyCreateInstance).toBeTruthy();
		});
		jest.spyOn(console, "warn").mockImplementation(() => {});
		const { asFragment } = renderWithRouter(
			<Suspense fallback="">
				<App />
			</Suspense>
		);

		await waitFor(() => {
			expect(asFragment()).toMatchSnapshot();
		});
	});
});

describe("#AppTest", () => {
	it.skip("handles device rotations", async () => {
		const spyConfig = jest.spyOn(localforage, "config");
		const spyCreateInstance = jest.spyOn(localforage, "createInstance");
		// jest.spyOn(window, "addEventListener").mockImplementation(() => {});

		await waitFor(async () => {
			expect(spyConfig).toBeTruthy();
			expect(spyCreateInstance).toBeTruthy();
		});

		renderWithRouter(
			<Suspense fallback="">
				<App />
			</Suspense>,
			{ route: "/home" }
		);

		// expect(window.addEventListener).toHaveBeenCalledWith(< String, typeof Function, typeof {} > "");
		resizeWindow(500, 800);
		const container = screen.getByTestId("homeWrapper");
		await waitFor(() => {
			expect(container).toBeInTheDocument();
		});
	});
});
