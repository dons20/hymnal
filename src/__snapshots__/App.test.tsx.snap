// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#App renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="root css-0"
  >
    <section
      class="app_body css-0"
    >
      <main
        class="app_inner  css-1157erc"
      >
        <div
          class="wrapper"
        />
      </main>
    </section>
  </div>
</DocumentFragment>
`;
